src/frontend/src/optimizer/plan_node/logical_hop_window.rs

```diff
@@ -65,33 +66,58 @@ impl LogicalHopWindow {
+            .iter()
+            .position(|&idx| idx == input.schema().len());
+        let window_end_index = output_indices
+            .iter()
+            .position(|&idx| idx == input.schema().len() + 1);
         let pk_indices = (|| {
             let input_pk = input
                 .pk_indices()
                 .iter()
                 .filter_map(|&pk_idx| output_indices.iter().position(|&idx| idx == pk_idx));
-            let window_pk = if output_indices.contains(&input.schema().len()) {
-                std::iter::once(
-                    output_indices
-                        .iter()
-                        .position(|&idx| idx == input.schema().len())
-                        .unwrap(),
-                )
-            } else if output_indices.contains(&(input.schema().len() + 1)) {
-                std::iter::once(
-                    output_indices
-                        .iter()
-                        .position(|&idx| idx == input.schema().len() + 1)
-                        .unwrap(),
-                )
+            let window_pk = if let Some(start_idx) = window_start_index {
+                std::iter::once(start_idx)
+            } else if let Some(end_idx) = window_end_index {
+                std::iter::once(end_idx)
             } else {
                 // If neither `window_start` or `window_end` is in `output_indices`, pk cannot be
                 // derived. In this situation, return empty vec.
                 return Vec::new();
             };
             input_pk.chain(window_pk).collect_vec()
         })();
-        let base = PlanBase::new_logical(ctx, actual_schema, pk_indices);
+        let functional_dependency = {
+            let input_fd =
+                ColIndexMapping::identity_or_none(input.schema().len(), original_schema.len())
+                    .composite(&ColIndexMapping::with_remaining_columns(
+                        &output_indices,
+                        original_schema.len(),
+                    ))
+                    .rewrite_functional_dependency_set(input.functional_dependency().clone());
+            let mut current_fd = FunctionalDependencySet::new(actual_schema.len());
+            for fd in input_fd.as_dependencies() {
+                if let Some(start_idx) = window_start_index {
+                    let mut fd = fd.clone();
+                    fd.set_from(start_idx, true);
+                    current_fd.add_functional_dependency(fd);
+                }
+                if let Some(end_idx) = window_end_index {
+                    let mut fd = fd.clone();
+                    fd.set_from(end_idx, true);
+                    current_fd.add_functional_dependency(fd);
+                }
+            }
+            if let Some(start_idx) = window_start_index {
+                current_fd.add_key(&[start_idx])
+            }
+            if let Some(end_idx) = window_end_index {
+                current_fd.add_key(&[end_idx])
+            }
```

Comment:
window start & end themselfs are not keys. should include input key (i.e., window_pk?)